packages/file-types/zarr/src/anndata-loaders/ObsFeatureMatrixAnndataLoader.js

```diff
@@ -10,6 +10,29 @@ import {
 // value returned from the load function.
 const toObject = data => ({ data });
 
+const filterOutEveryOther = (arr) => {
+  const filtered = new arr.constructor(arr.length / 2);
+  for (let i = 0; i < (arr.length / 2); i++) {
+    filtered[i] = arr[i * 2];
+  }
+  return filtered;
+};
+
+const castByBuffer = (arr) => {
+  const buffer = new ArrayBuffer(arr.length * 4)
+  const filtered = new DataView();
+  const viewOfArr = new DataView(arr.buffer)
+  for (let i = 0; i < arr.length; i++) {
+    filtered.setInt32(i * 4, viewOfArr.getInt32(i * 8, True));
+  }
+  return new Int32Array(buffer);
+};
+
+// eslint-disable-next-line no-undef
+const maybeDowncastInt64 = data => (data.constructor === BigInt64Array
+  ? filterOutEveryOther(new Int32Array(data.buffer))
+  : data);
+
```

Comment:
@manzt Are you aware of a faster or better way of going from BigInt64Array to Int32Array? Some comparisons and alternatives:

```javascript
const filterOutEveryOther = (arr) => {
  const filtered = new arr.constructor(arr.length / 2);
  for (let i = 0; i < (arr.length / 2); i++) {
    filtered[i] = arr[i * 2];
  }
  return filtered;
};

const castByBuffer = (arr) => {
  const buffer = new ArrayBuffer(arr.length * 4)
  const filtered = new DataView(buffer);
  const viewOfArr = new DataView(arr.buffer)
  for (let i = 0; i < arr.length; i++) {
    filtered.setInt32(i * 4, viewOfArr.getInt32(i * 8, false));
  }
  return new Int32Array(buffer);
};

const castByArr = (arr) => {
  const filtered = new Int32Array(arr.length);
  for (let i = 0; i < arr.length; i++) {
    filtered[i] = Number(arr[i])
  }
  return filtered;
};


let data = new BigInt64Array([1n, 3n, 10n, 15n])
console.log('ops produce equal results?', castByArr(data), castByBuffer(data), filterOutEveryOther(new Int32Array(data.buffer)))

data = BigInt64Array.from(Array.from(Array(10000000).keys()).map(BigInt))
let t = performance.now()
filterOutEveryOther(new Int32Array(data.buffer))
console.log('filtering took ', performance.now() - t)

t = performance.now()
castByBuffer(data)
console.log('cast by buffer took ', performance.now() - t)

t = performance.now()
castByArr(data)
console.log('cast by arr took ', performance.now() - t)
```

Reply:
Filtering _seems_ to be fastest, if still most complicated

Reply:
Sorry this is outdated, didn't notice the `castByBuffer` there.  This PR uses the filtering op:
```javascript
const maybeDowncastInt64 = data => (data.constructor === BigInt64Array
  ? filterOutEveryOther(new Int32Array(data.buffer))
  : data);
```

Reply:
If we can safely ignore the the upper bits, I think this would be preferable.

```ts
/**
 * Fast conversion of a BigInt64Array to an Int32Array.
 *
 * Takes the lower 32 bits of each 64-bit value in the BigInt64Array.
 * Assumes the upper 32 bits are not used and little-endian byte order.
 *
 * @param arr - The little-endian BigInt64Array to convert.
 * @returns A Int32Array containing the lower 32 bits of each 64-bit value.
 */
function convertBigInt64ArrayToInt32Array(arr: BigInt64Array): Int32Array {
	if (!(arr instanceof BigInt64Array)) {
		throw new Error("Expected a BigInt64Array");
	}
	let out = new Int32Array(arr.length);
	// Create a view of the BigInt64Array buffer as a Int32Array (2x elements of out)
	let view = new Int32Array(arr.buffer);
	for (let i = 0; i < view.length; i++) {
		/**
		 * Get the lower 32 bits of each 64-bit value.
		 *
		 * Since each 64-bit value takes up 2 slots in the Int32Array view, we
		 * multiply the index by 2 to get the correct position (assuming
		 * little-endian).
		 */
		out[i] = view[i * 2];
	}
	return out;
}
```

Reply:
Thanks @manzt I think I will stick with what we have.  The operation doesn't take that long but `filter` is definitely much slower and these matrices can get big so would rather not take the risk.  Thanks!

Reply:
Ok no worries. I would however still recommend asserting the upper bits are 0, even if there is a slight performance cost. Also, rather than a generic function I'd probably make a specialized function that is well documented about the behavior (`(BigInt64Array) -> Int32Array`). There's a lot happening in the line where the `Int32Array` view over the BigInt buffer is created and `filterOutEveryOther` is called (and it's within a ternary). I couldn't see that bit of code and `convertBigInt64ArrayToInt32Array` are essentially identical implementations.

Also, I benchmarked all the methods, including a checked version of the `convertBigInt64ArrayToInt32Array` and it's 2x slower from the unchecked version but still faster than all other options. 

```
❯ deno bench bench.ts
cpu: Apple M3 Max
runtime: deno 1.41.3 (aarch64-apple-darwin)

file:///Users/manzt/github/manzt/chromoload/bench.ts
benchmark                                      time (avg)        iter/s             (min … max)       p75       p99      p995
----------------------------------------------------------------------------------------------- -----------------------------
filterOutEveryOther                             6.62 ms/iter         151.2     (6.45 ms … 7.67 ms) 6.62 ms 7.67 ms 7.67 ms
castByBuffer                                   13.96 ms/iter          71.6   (13.84 ms … 14.32 ms) 13.98 ms 14.32 ms 14.32 ms
castByArr                                       83.5 ms/iter          12.0   (82.03 ms … 85.75 ms) 84.21 ms 85.75 ms 85.75 ms
convertBigInt64ArrayToInt32ArrayChecked        17.05 ms/iter          58.7   (16.78 ms … 17.91 ms) 17.17 ms 17.91 ms 17.91 ms
convertBigInt64ArrayToInt32ArrayUnchecked       6.57 ms/iter         152.2     (6.45 ms … 7.61 ms) 6.62 ms 7.61 ms 7.61 ms

summary
  filterOutEveryOther
   1.01x slower than convertBigInt64ArrayToInt32ArrayUnchecked
   2.11x faster than castByBuffer
   2.58x faster than convertBigInt64ArrayToInt32ArrayChecked
   12.62x faster than castByArr
```

<details>
   <summary>bench.ts</summary>

```js
import { assertEquals } from "jsr:@std/assert";

const filterOutEveryOther = (arr: any) => {
  const filtered = new arr.constructor(arr.length / 2);
  for (let i = 0; i < (arr.length / 2); i++) {
    filtered[i] = arr[i * 2];
  }
  return filtered;
};

const castByBuffer = (arr: any) => {
  const buffer = new ArrayBuffer(arr.length * 4);
  const filtered = new DataView(buffer);
  const viewOfArr = new DataView(arr.buffer);
  for (let i = 0; i < arr.length; i++) {
    filtered.setInt32(i * 4, viewOfArr.getInt32(i * 8, false));
  }
  return new Int32Array(buffer);
};

const castByArr = (arr: any) => {
  const filtered = new Int32Array(arr.length);
  for (let i = 0; i < arr.length; i++) {
    filtered[i] = Number(arr[i]);
  }
  return filtered;
};

/**
 * Fast conversion of a BigInt64Array to an Int32Array.
 *
 * Takes the lower 32 bits of each 64-bit value in the BigInt64Array.
 * Assumes the upper 32 bits are not used and little-endian byte order.
 *
 * @param arr - The little-endian BigInt64Array to convert.
 * @returns A Int32Array containing the lower 32 bits of each 64-bit value.
 */
function convertBigInt64ArrayToInt32ArrayChecked(
  arr: BigInt64Array,
): Int32Array {
  if (!(arr instanceof BigInt64Array)) {
    throw new Error("Expected a BigInt64Array");
  }
  let out = new Int32Array(arr.length);
  // Create a view of the BigInt64Array buffer as a Int32Array (2x elements of out)
  let view = new Int32Array(arr.buffer);
  for (let i = 0; i < arr.length; i++) {
    if (view[i * 2 + 1] !== 0) {
      throw new Error("Upper 32 bits are not zero");
    }
    /**
     * Get the lower 32 bits of each 64-bit value.
     *
     * Since each 64-bit value takes up 2 slots in the Int32Array view, we
     * multiply the index by 2 to get the correct position (assuming
     * little-endian).
     */
    out[i] = view[i * 2];
  }
  return out;
}

/**
 * Fast conversion of a BigInt64Array to an Int32Array.
 *
 * Takes the lower 32 bits of each 64-bit value in the BigInt64Array.
 * Assumes the upper 32 bits are not used and little-endian byte order.
 *
 * @param arr - The little-endian BigInt64Array to convert.
 * @returns A Int32Array containing the lower 32 bits of each 64-bit value.
 */
function convertBigInt64ArrayToInt32ArrayUnchecked(
  arr: BigInt64Array,
): Int32Array {
  if (!(arr instanceof BigInt64Array)) {
    throw new Error("Expected a BigInt64Array");
  }
  let out = new Int32Array(arr.length);
  // Create a view of the BigInt64Array buffer as a Int32Array (2x elements of out)
  let view = new Int32Array(arr.buffer);
  for (let i = 0; i < arr.length; i++) {
    /**
     * Get the lower 32 bits of each 64-bit value.
     *
     * Since each 64-bit value takes up 2 slots in the Int32Array view, we
     * multiply the index by 2 to get the correct position (assuming
     * little-endian).
     */
    out[i] = view[i * 2];
  }
  return out;
}

let data = BigInt64Array.from(Array.from(Array(10_000_000).keys()).map(BigInt));

Deno.bench("filterOutEveryOther", { baseline: true }, () => {
  let result = filterOutEveryOther(new Int32Array(data.buffer));
  assertEquals(result.length, data.length);
});

Deno.bench("castByBuffer", () => {
  let result = castByBuffer(data);
  assertEquals(result.length, data.length);
});

Deno.bench("castByArr", () => {
  let result = castByArr(data);
  assertEquals(result.length, data.length);
});

Deno.bench("convertBigInt64ArrayToInt32ArrayChecked", () => {
  let result = convertBigInt64ArrayToInt32ArrayChecked(data);
  assertEquals(result.length, data.length);
});

Deno.bench("convertBigInt64ArrayToInt32ArrayUnchecked", () => {
  let result = convertBigInt64ArrayToInt32ArrayUnchecked(data);
  assertEquals(result.length, data.length);
});
```

</details>

Reply:
> I would however still recommend asserting the upper bits are 0, even if there is a slight performance cost.

I had spoken to Isaac about this issue and we agreed there's no chance of ever hitting this problem because this is just the data from the matrix (for example we will never see more than 2^31 - 1 transcripts for a given cell - Isaac pointed out that he too has only ever seen this data type in testing).   The `indptr` or `indices` of a sparse matrix definitely can and will go over that limit as they index, but for data I am confident we will never have a problem.  I really don't want to add more time on to this processing.

I'll update the function for sure though! Thanks for the implementation!

Reply:
Ok nice! Maybe good to leave a comment in the function why we can safely ignore the upper bits. Or add the `Unchecked` suffix for clarity.